File: 213. House Robber II.cpp
class Solution {
public:
    int rob(vector<int>& nums) {
        if(nums.size() == 0){
            return 0;
        }
        if(nums.size() == 1){
            return nums[0];
        }
        int n = nums.size();
        vector<int> dp1(n + 1, 0);
        vector<int> dp2(n + 1, 0);
        dp1[0] = 0;
        dp1[1] = nums[0];
        dp2[0] = 0;
        dp2[1] = 0;
        dp2[2] = nums[1];
        for(int i = 2; i < n; i++) {
            dp1[i] = max(dp1[i - 1],dp1[i - 2] + nums[i - 1]);
        }
        for(int i = 3; i <= n; i++) {
            dp2[i] = max(dp2[i - 1],dp2[i - 2] + nums[i - 1]);
        }
        return max(dp1[n - 1],dp2[n]);
    }
};

File: 213. House Robber II.java
class Solution {
    public int rob(int[] nums) {
        int len = nums.length;
        if(len == 1) {
            return nums[0];
        };
        
        int result = 0;
        int pre, cur, temp;
        pre = 0;
        cur = 0;
        for(int i = 0; i < len - 1; i++) {
            temp = cur;
            cur = Math.max(pre + nums[i], cur);
            pre = temp;
        }
        result = Math.max(result, cur);        
        pre = 0;
        cur = 0;
        for(int i = 1; i < len; i++) {
            temp = cur;
            cur = Math.max(pre + nums[i], cur);
            pre = temp;
        }
        return Math.max(result, cur);
    }
}